Settings > Organization Settings: Code revamp

RachelElysia opened the pull request:
Cerra #2477 , Cerra #3487

Part of #3516 

- Convert javascript to typeScript
- Convert class components to functional components following new patterns (e.g. useState hooks, onChange functions)
- Convert AppConfigForm to use Context API instead of Redux for loadAll and update calls
- Update config interface accordingly (e.g. not flatten and unflatten through helpers)
- Convert validators to follow new config interface
- e2e test updated and tests agreed upon config flows (26 form fields are tested in settingsflow.spec.ts)


Updates since loom: Renamed helpers.ts to constants.ts and moved two data previews into constants.ts
https://www.loom.com/share/1923a0c01b264a7c959cd9cd38109982

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

- [x] Changes file added (for user-visible changes)
- [x] Added/updated tests
- [x] Manual QA for all new/changed functionality
- [x] QA process: Test update call to all 33 form fields and AppConfigForm validations lines 158-221

lukeheath commented on frontend/components/forms/fields/InputField/InputField.jsx:
```diff
@@ -36,6 +37,7 @@ class InputField extends Component {
     onFocus: noop,
     type: "text",
     value: "",
+    target: false,
```
Can we change this prop name to something that answers true/false? Like `includeTarget` or `parseTarget`.

lukeheath commented on frontend/components/forms/admin/AppConfigForm/AppConfigFormFunctional.tsx:
```diff
@@ -0,0 +1,1121 @@
+
+  // █▀▀ █▀█ █▀█ █▀▄▀█   █▀ ▀█▀ ▄▀█ ▀█▀ █▀▀
+  // █▀░ █▄█ █▀▄ █░▀░█   ▄█ ░█░ █▀█ ░█░ ██▄
+  const [iterateFormData, setIterateFormData] = useState<any>({
```
The name `iterateFormData` is confusing to me. I'm not sure why iterate is including. Could it be just `formData`?

lukeheath replied:
This object should also be an interface so you don't have to use `any`.

lukeheath commented on frontend/components/forms/admin/AppConfigForm/AppConfigFormFunctional.tsx:
```diff
@@ -0,0 +1,1121 @@
+    disableLiveQuery: formData.server_settings.live_query_disabled || false,
+  });
+
+  const {
+    orgName,
+    orgLogoURL,
+    serverURL,
+    enableSSO,
+    idpName,
+    entityID,
+    issuerURI,
+    idpImageURL,
+    metadata,
+    metadataURL,
+    enableSSOIDPLogin,
+    enableSMTP,
+    smtpSenderAddress,
+    smtpServer,
+    smtpPort,
+    smtpEnableSSLTLS,
+    smtpAuthenticationType,
+    smtpUsername,
+    smtpPassword,
+    smtpAuthenticationMethod,
+    agentOptions,
+    enableHostStatusWebhook,
+    hostStatusWebhookDestinationURL,
+    hostStatusWebhookHostPercentage,
+    hostStatusWebhookDaysCount,
+    enableUsageStatistics,
+    domain,
+    verifySSLCerts,
+    enableStartTLS,
+    enableHostExpiry,
+    hostExpiryWindow,
+    disableLiveQuery,
+  } = iterateFormData;
```
Right now these are all destructured to type `any` because of the `useState`. If you provide an interface to `useState` all of these will destructure to the correct types, which should make  typescript a lot more useful in this file.

lukeheath commented on frontend/components/forms/admin/AppConfigForm/AppConfigFormFunctional.tsx:
```diff
@@ -0,0 +1,1121 @@
+    // Validators
+    const errors: any = {};
+
+    if (enableSSO) {
```
`useReducer` is the newer pattern for form validation. But, that pattern is only implemented in cases where we have a context object (app, query, policy). I'm looking into how we want to implement it here, but I'll create a separate ticket for that. For our purposes, this validation works and is totally fine for this ticket.

lukeheath replied:
I noticed this same code in `validate.js` that the old `AppConfigForm` was using. Can we keep using the imported file for validation to cut down on line size in this component?

RachelElysia replied:
As I mentioned in my video, we'd have to pass formData into validate everywhere and then destructure on a separate file which ends up adding a lot of extra lines of code everywhere. I cleaned this file down to 1078 lines from 1300 which is great!

lukeheath commented on frontend/components/forms/admin/AppConfigForm/AppConfigFormFunctional.tsx:
```diff
@@ -0,0 +1,1121 @@
+    }
+
+    // formDataToSubmit mirrors formatting of API not UI
+    const formDataToSubmit = {
```
This is painful but unavoidable right now. Previously this was being rebuilt inside of `validate.js`. Can we do that again to cut down on line size?

lukeheath commented on frontend/components/forms/admin/AppConfigForm/AppConfigFormFunctional.tsx:
```diff
@@ -0,0 +1,1121 @@
+          <InputField
+            label="Organization name"
+            onChange={handleInputChange}
+            target
```
change to `parseTarget`?

lukeheath commented on frontend/components/forms/fields/Dropdown/Dropdown.jsx:
```diff
@@ -56,14 +58,26 @@ class Dropdown extends Component {
+    console.log("name", name);
+    console.log("selected.value", selected.value);
+    if (target) {
+      if (clearable && selected === null) {
+        onChange(name, null);
+      } else {
+        onChange(name, selected.value);
+      }
     } else {
-      onChange(selected.value);
+      if (clearable && selected === null) {
+        onChange(null);
+      } else if (multi) {
+        onChange(selected.map((obj) => obj.value).join(","));
+      } else {
+        onChange(selected.value);
+      }
```
This block is getting a little if-fy. One trick is using `return` to short circuit the function like this: 

```javascript
    if (target) {
      clearable && selected === null
        ? onChange(name, null)
        : onChange(name, selected.value);
      return;
    }

    if (clearable && selected === null) {
      onChange(null);
      return;
    }

    if (multi) {
      onChange(selected.map((obj) => obj.value).join(","));
      return;
    }

    onChange(selected.value);
```

RachelElysia commented:
@lukeheath - We need to look into redux and check where config is being used from Redux and make sure that it's updated as needed.

lukeheath reviewed (changes requested):
After testing the Redux state it appears that config changes are not being applied to Redux until the window is reloaded. 

I tested this by adding `console.log("config: ", state.app.config);` to line 199 of `CoreLayout.jsx`. 

Referencing `App.tsx`, it appears we need to pass the updated config to a `setConfig()` function on the `AppContext`: 

```typescript
const {setConfig} = useContext(AppContext);
setConfig(updatedConfig);
```

We'd want to the pass the entire config object we get back after making a PATCH to `/config`.

RachelElysia commented:
@lukeheath 
Added lines 68-73 on AppSettingsPage.tsx to update config in both Redux (still being used in 3 places) and AppContext.
```
          // Config must be updated in both Redux and AppContext
          dispatch(getConfig())
            .then((config: IConfig) => {
              setConfig(config);
            })
            .catch(() => false);
```

Rebased off main to resolve conflicts, made sure new changes from conflict were implemented (blockAutocomplete parameter was on username/password).

lukeheath reviewed (approved):
Looks great!
